Sources/Testing/Parameterization/CustomTestArgument.swift

```diff
@@ -0,0 +1,154 @@
+#if canImport(Foundation)
+    guard Configuration.current?.isTestArgumentEncodingEnabled ?? false else {
+      return nil
+    }
+
+    func _customArgumentWrapper(for value: some CustomTestArgumentEncodable) -> some Encodable {
+      _CustomArgumentWrapper(value: value)
+    }
+
+    let encodableValue: (any Encodable)? = if let customEncodable = value as? any CustomTestArgumentEncodable {
+      _customArgumentWrapper(for: customEncodable)
+    } else if let encodable = value as? any Encodable {
+      encodable
+    } else if let identifiable = value as? any Identifiable, let encodableID = identifiable.id as? any Encodable {
+      encodableID
+    } else {
+      nil
+    }
+
+    guard let encodableValue else {
+      return nil
+    }
+
+    self = .init(bytes: try Self._encode(encodableValue, parameter: parameter))
+#else
+    nil
+#endif
+  }
+
+#if canImport(Foundation)
+  /// Encode the specified test argument value and store its encoded
+  /// representation as an array of bytes suitable for storing in an instance of
+  /// ``Test/Case/Argument/ID-swift.struct``.
+  ///
+  /// - Parameters:
+  ///   - value: The value to encode.
+  ///   - parameter: The parameter of the test function to which this argument
+  ///     value was passed.
+  ///
+  /// - Returns: An array of bytes containing the encoded representation.
+  ///
+  /// - Throws: Any error encountered during encoding.
+  private static func _encode(_ value: some Encodable, parameter: Test.ParameterInfo) throws -> [UInt8] {
+    let encoder = JSONEncoder()
+
+    // Keys must be sorted to ensure deterministic matching of encoded data.
+    encoder.outputFormatting.insert(.sortedKeys)
+
+    // Set user info keys which clients may wish to use during encoding.
+    encoder.userInfo[._testParameterUserInfoKey] = parameter
+
+    return .init(try encoder.encode(value))
+  }
+#endif
+}
+
+/// A encodable type which wraps a ``CustomTestArgumentEncodable`` value.
+private struct _CustomArgumentWrapper<T>: Encodable where T: CustomTestArgumentEncodable {
+  /// The value this instance wraps, which implements custom test argument
+  /// encoding logic.
+  var value: T
+
+  func encode(to encoder: any Encoder) throws {
+    try value.encodeTestArgument(to: encoder)
+  }
+}
+
+// MARK: - Additional coding user info
+
+extension CodingUserInfoKey {
+  /// A coding user info key whose value is a ``Test/ParameterInfo``.
+  fileprivate static var _testParameterUserInfoKey: Self {
+    Self(rawValue: "org.swift.testing.coding-user-info-key.parameter")!
+  }
+}
+
+extension Encoder {
+  /// The test parameter which the test argument being encoded was passed to, if
+  /// any.
+  ///
+  /// The value of this property is non-`nil` when this encoder is being used to
+  /// encode an argument passed to a parameterized test function.
+  @_spi(ExperimentalParameterizedTesting)
+  public var parameter: Test.ParameterInfo? {
```

Comment:
Since we're extending a stdlib type here, we should pick a less generic property name. How about `testParameterInfo`?

Reply:
I'll go with `testParameter`, since elsewhere we just call it "parameter" (despite its formal type including the word "Info")